### PR TITLE
Chore/add flow to tracking context file

### PR DESCRIPTION
**Pull request by colincclark**
- Removed unrequired global flow config file
- Added flow to the tracking package
- Added actual flow code, partially, to the tracking context
- complete flow code for tracking context then get review

### CODE REVIEW

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/bf4308e97bc1579e1b8a52a8d503ba676ba09934/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/f49b119f41e86981ce2239f15549742b5affa3bf/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/4646eedc14fa8c1ed3ee5320c2de42bf637cb93c/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/32b5863726a18473a928b40bbf5358eb74cea837/dextrose-test/index.html

File: packages/tracking/tracking-context.js [hunk 1]
```diff
@@ -1,19 +1,35 @@
...
 import trackingContextTypes from "./tracking-context-types";
 import resolveAttrs from "./resolve-attrs";
 
+type TrackingProps = {
+  analyticsStream: () => mixed
+};
+
+type TrackingObject = {
+  getAttrs: () => mixed,
+  trackingObject: any,
```
Comment (toboid) [inline packages/tracking/tracking-context.js]: `trackingObject` should always be a string

File: packages/tracking/tracking-context.js [hunk 2]
```diff
@@ -1,19 +1,35 @@
...
 import hoistNonReactStatic from "hoist-non-react-statics";
 import trackingContextTypes from "./tracking-context-types";
 import resolveAttrs from "./resolve-attrs";
 
+type TrackingProps = {
+  analyticsStream: () => mixed
+};
+
+type TrackingObject = {
+  getAttrs: () => mixed,
```
Comment (toboid) [inline packages/tracking/tracking-context.js]: `getAttrs` should return an object

File: packages/tracking/tracking-context.js [hunk 3]
```diff
@@ -1,19 +1,35 @@
...
+import type { Component } from "react";
 import PropTypes from "prop-types";
 import getDisplayName from "react-display-name";
 import _get from "lodash.get";
 import hoistNonReactStatic from "hoist-non-react-statics";
 import trackingContextTypes from "./tracking-context-types";
 import resolveAttrs from "./resolve-attrs";
 
+type TrackingProps = {
+  analyticsStream: () => mixed
```
Comment (toboid) [inline packages/tracking/tracking-context.js]: `analyticsStream` doesn't return anything at the moment

File: packages/tracking/tracking-context.js [hunk 4]
```diff
@@ -1,19 +1,35 @@
...
+};
+
+type TrackingObject = {
```
Comment (toboid) [inline packages/tracking/tracking-context.js]: I think the name `TrackingObject` might be a bit confusing as also contains a `trackingObject` field. I guess this type really represents  `TrackingContextArguments`?

File: packages/tracking/tracking-context.js [hunk 5]
```diff
@@ -1,19 +1,35 @@
...
+import React from "react";
+import type { Component } from "react";
 import PropTypes from "prop-types";
 import getDisplayName from "react-display-name";
 import _get from "lodash.get";
 import hoistNonReactStatic from "hoist-non-react-statics";
 import trackingContextTypes from "./tracking-context-types";
 import resolveAttrs from "./resolve-attrs";
 
+type TrackingProps = {
```
Comment (toboid) [inline packages/tracking/tracking-context.js]: Tiny think but how would you feel about calling this `TrackingContextProps`?

File: packages/tracking/package.json [hunk 6]
```diff
@@ -39,13 +39,10 @@
     "enzyme": "3.3.0",
     "enzyme-adapter-react-16": "1.1.0",
     "eslint": "4.9.0",
+    "flow-bin": "0.63.1",
     "jest": "21.2.1",
     "prettier": "1.8.2",
     "raf": "3.4.0",
-    "react": "16.2.0",
```
Comment (toboid) [inline packages/tracking/package.json]: There's been a few discussions about what dependencies should go where. Probably worth staying consistent with the other packages for now and change all together if there is an agreement to do so?

Comment (colincclark) [reply packages/tracking/package.json]: flow-types do not work with packages declared twice in a package.json file

Comment (toboid) [reply packages/tracking/package.json]: Ah interesting

**Comment by nikhedonia** [comment]
https://github.com/newsuk/times-components/blob/4eecd9126ae2982a85df6c7e75f18fe9bdc7aecf/packages/tracking/tracking-context.js#L132

flowtype will probably complain as neither WithTrackingContext nor WrappedComponent defines the existance of this field.

You could  add `static defaultProps : mixed` to the class definition and use sth. like 
 `= (WrappedComponent.defaultProps||{})`
But I'm not sure what our styleguide says about this practice.

**Comment by colincclark** [comment]
This is now waiting for https://github.com/newsuk/times-components/pull/571

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/7b164f7d0f845f7a7bf2e4a7e5da2d9021804062/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/472fdd17c021e226efe75b09f83588ff4f7002e4/dextrose-test/index.html

File: packages/tracking/tracking-context.js [hunk 7]
```diff
@@ -1,26 +1,55 @@
...
 import PropTypes from "prop-types";
 import getDisplayName from "react-display-name";
-import _get from "lodash.get";
+import get from "lodash.get";
 import hoistNonReactStatic from "hoist-non-react-statics";
 import trackingContextTypes from "./tracking-context-types";
 import resolveAttrs from "./resolve-attrs";
 
+type AnalyticsEventObjectType = {
+  object: {},
```
Comment (toboid) [inline packages/tracking/tracking-context.js]: I think `object` should always be a string but may not always be present.

File: packages/tracking/tracking-context.js [hunk 8]
```diff
@@ -87,9 +125,9 @@ const withTrackingContext = (
       return !this.context || !this.context.tracking;
     }
 
-    analyticsStream(...args) {
+    analyticsStream(...args: Array<any>) {
```
Comment (toboid) [inline packages/tracking/tracking-context.js]: I think this arg will always be a single event at the moment as there is only one call (at line 121)

File: packages/tracking/tracking-context.js (hunk 7 above)
Comment (toboid) [reply packages/tracking/tracking-context.js]: It's possibly not the easiest to model in flow. I think the rule is that the root context in the component hierarchy (which is determined at runtime by the presence of a parent react context) must emit events with an `object` field but any child contexts can omit that field and expect it to be added by the root context.

File: packages/tracking/tracking-context.js (hunk 8 above)
Comment (colincclark) [reply packages/tracking/tracking-context.js]: @toboid I tried this, but I kept getting `Expected array`

File: packages/tracking/tracking-context.js (hunk 7 above)
Comment (colincclark) [reply packages/tracking/tracking-context.js]: Doesn't line 114 mean the `decoratedEvent.object` can either be an object or a string? `object` is always an object, and `trackingObjectName` is always a string, no?

File: packages/tracking/tracking-context.js (hunk 8 above)
Comment (colincclark) [reply packages/tracking/tracking-context.js]: (because of the rest operator)

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/86e06f421f78c99241c87dec0bff9ff645be25ae/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/e560678276d804e7b8b1b2e04db4c83beff3e2bd/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/069efbcdfad00c796fa9074661663fde31ce0837/dextrose-test/index.html

**Comment by colincclark** [comment]
@craigbilner looking good!

**Pull request merged**